src/we_aes_gcm.c

```diff
@@ -356,153 +302,142 @@ static int we_aes_gcm_cipher(EVP_CIPHER_CTX *ctx, unsigned char *out,
+                    ret = (int)len;
                 }
-                /* No data decrypted, yet, so return 0. */
-                ret = 0;
-            }
-            else
-        #endif
-            {
-                ret = we_aes_gcm_decrypt(aes, out, in, len);
             }
-        }
-
-    #ifdef WE_AES_GCM_DECRYPT_ON_FINAL
-        if (aes->enc || (!aes->enc && aes->tagLen > 0))
-    #endif
-        {
-            /* Dispose of any AAD - all used now. */
-            OPENSSL_free(aes->aad);
-            aes->aad = NULL;
-            aes->aadLen = 0;
-        }
-        if (ret == 1) {
-            if (aes->ivInc) {
-                int i;
-                for (i = aes->ivLen - 1; i >= aes->ivLen - 8; i--) {
-                    if ((++aes->iv[i]) != 0) {
-                        break;
+            else {
+                if (!aes->ivSet) {
+                    /* Set extern IV. */
+                    WOLFENGINE_MSG(WE_LOG_CIPHER, "Setting external IV");
+                    rc = wc_AesGcmSetExtIV(&aes->aes, aes->iv, aes->ivLen);
+                    if (rc != 0) {
+                        WOLFENGINE_ERROR_FUNC(WE_LOG_CIPHER,
+                            "wc_AesGcmSetExtIV", rc);
+                        ret = -1;
+                    }
+                    else {
+                        aes->ivSet = 1;
                     }
                 }
-                aes->ivInc = 0;
-                aes->ivSet = 0;
-            }
-            ret = (int)len;
-        }
-    }
-    else if ((ret == 1) && (len == 0)) {
-    #ifdef WE_AES_GCM_DECRYPT_ON_FINAL
-        if (!aes->enc) {
-            ret = we_aes_gcm_decrypt(aes, out, aes->decryptBuf,
-                aes->decryptBufLen);
-            aes->tagLen = 0;
-        }
-    #endif
+                if (ret == 1 && (len != 0 || in != NULL)) {
+                    /* "Update" called. */
+                    if (len != 0 && in != NULL) {
+                        aes->tmp = (unsigned char*)OPENSSL_malloc(len);
+                        if (aes->tmp == NULL) {
+                            WOLFENGINE_ERROR_FUNC_NULL(WE_LOG_CIPHER,
+                                "OPENSSL_malloc", aes->tmp);
+                            ret = -1;
+                        }
+                        else {
+                            /* 
+                             * All encryption/decryption is deferred to the
+                             * "final" call. We just save the input data and
+                             * the address of the output buffer on "update"
+                             * calls.
+                             */
+                            XMEMCPY(aes->tmp, in, len);
+                            aes->tmpLen = (int)len;
+                            aes->outputBuf = out;
+                            /* Return length of buffered input data. */
+                            ret = (int)len;
+                        }
+                    }
+                    else {
```

Comment:
Is len != 0 and in == NULL an error?
Does it not ever happen?

Reply:
I think it should be treated as an error. Will fix.